keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,436 @@
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `Local` policy for `service.spec.externalTrafficPolicy` and
+`service.spec.internalTrafficPolicy` allows configuring services so
+that they will always deliver to local endpoints, avoiding an
+unnecessary hop to another node when the request could just as easily
+be processed locally.
+
+However, they both fail entirely if there are no endpoints for the
+service on the node. This tradeoff is unacceptable for some use cases.
+
+The [Topology Aware Hints feature] provides a way of signaling to
+kube-proxy that it should optimize service delivery on a
+per-availability zone basis, but does not provide any way to optimize
+on any other basis.
+
+This KEP adds a new topology hint, to tell kube-proxy that a Service
+is expected to have an endpoint on every node most of the time, and so
```

Comment:
I was vague here because it's not fully figured out below yet.

I talk below about _kube-controller-manager_ deriving it, but not kube-proxy, because kube-proxy doesn't watch Pods, so it has less information. I guess kube-proxy could still validate "has endpoints on every node / almost every node", but it can't validate "was created by a DaemonSet". (But maybe we don't care about that? The criteria for deciding when to use this are also undecided.)

I guess one nice thing about having the EndpointSlice controller write to an actual EndpointSlice field is that it makes it very clear when the feature is being used, whereas if kube-proxy decided by itself, you wouldn't be able to tell from looking at the Service/Endpoints.

Reply:
I agree that a central controller in kube-controller-manager is the best place to determine if this feature should be enabled. I'm just hoping we can find a way to indicate that as centrally as possible for a Service so we're not updating every individual endpoint whenever this changes.

Reply:
I remember with zone-based hints, one of the biggest concerns was to avoid flapping between an enabled and disabled state (requiring updates to all relevant EPS). I think whatever we do here should have some kind of detail written about how we'd avoid that.

Reply:
yeah, some hysteresis, like: if < 90% of nodes have an endpoint, disable; if >= 95%, enable; if >= 90% && < 95%, leave it as it is set.